### ISSUE TITLE
Mouse wheel scrolling on tab bar does not scroll (visually)

### ISSUE THREAD

**Issue by v3ss0n** [atom/tabs#137]
Expected Behavior:

- scroll the tab strip visually 

Right now:

- When hidden , Only buffers are switched , tabs are not scrolled.

RESOLVED, to do horizontal scroll shift + mousewheel , thanks @izuzak .

**Comment by v3ss0n** [comment]
Also what you guys use to make animated screencaps

**Comment by izuzak** [comment]
> Also what you guys use to make animated screencaps

I use http://www.cockos.com/licecap/. 

Also, can you please keep issues focused on one problem or one feature request? Mixing several problems and feature requests into one issue makes it hard to understand what you want and track problems. For example, you could keep this issue focused on the request for the tab bar to auto scroll and open a new issue to report a problem with using the mouse to manually scroll the tab bar. Does that sound good?

**Comment by v3ss0n** [comment]
Thanks alot, not sure licecap works for linux , looking around.

> For example, you could keep this issue focused

Sounds good and i am following that.

#138 is for auto-scroll , this one is not behaving as intended.
while looking around , a comment on one-dark-ui Repo have a screencap with expected behavior for this issue :  

![](https://cloud.githubusercontent.com/assets/378023/7432018/2032ddf8-f05f-11e4-8292-74cb341544bf.gif)

I think i have too look into one-dark-ui for this ?

**Comment by v3ss0n** [comment]
Edited so we can focus more.

**Comment by izuzak** [comment]
@v3ss0n So, I just tried this with a mouse with a scroll wheel and it works for me.

The scroll wheel by default scrolls vertically, but what you need is horizontal scrolling (because the tab bar scrolls that way). The way I do that is that I press `shift` and while holding shift scroll the mouse wheel -- that triggers horizontal scrolling. Basically, you do the same thing you would do to scroll any web page horizontally with your mouse. (I use the same approach to scroll Atom's editor horizontally when soft wrap is turned off and I'm using a mouse with a scroll wheel). Does that work for you?

**Comment by v3ss0n** [comment]
Thanks a lot! now i am able to do it. I was forgetting shift + mouse wheel for many years. I think i only used it back in the days of windows LOL .

**Comment by izuzak** [comment]
Happy to hear that helps. I'm going to close this now then :smile:

**Comment by dantman** [comment]
Tabs only scroll horizontally. This is an editor, not a web page. Shift should not be required to scroll through the tabs.

**Comment by RamIdeas** [comment]
I'm with @dantman. Anywhere you see an area that scrolls only horizontally, vertical scrolling is mapped to the horizontal axis.

Plus, I have never even heard of using `shift` to scroll horizontally.

**Comment by v3ss0n** [comment]
As this one is closed , may be open a new issue?
i am agree with both of you.

**Comment by v3ss0n** [comment]
thanks a lot @mnquintana

**Comment by jacobmischka** [comment]
Scrolling over tabs to switch tabs (the current behavior) is pretty common. Works that way in every tab-based interface I can think of, Chrome, gnome-terminal, and zeal being three that I have open at the moment. Shift + scroll is also very common to scroll horizontally. The current behavior is good default behavior I think.

Maybe a setting might be nice to swap the two if this many people want it, but the current behavior should stay default. Or I think the shift + scroll behavior may just need to be advertised more, I didn't know about it until googling for it and coming across this issue.

**Comment by dessant** [comment]
@jacobmischka, i've just tested Chrome on Windows, if does not switch tabs upon scrolling.
In any case, Chrome does not support tab overscrolling, all tabs stay visible at all times.

Firefox might be a better example for the tab behavior people seem to expect.

I do agree that an option to toggle the default behavior might be the best route.

**Comment by jacobmischka** [comment]
Oh weird, it does for me on Linux. I didn't realize it was different. And yes it doesn't, you're right.

**Comment by madprops** [comment]
I naturally tried scrolling the tab bar when I noticed the tabs where hidden and there was no other UI element to control this. But it didn't work. Needing shift + scroll is unnatural, I had to come here to find out  that's how you do it.

**Comment by jacobmischka** [comment]
I still fully disagree that it should be changed, I just think the shift+scroll behavior needs to be described somewhere so people know about it.

**Comment by dessant** [comment]
@jacobmischka, but you agree about offering people an option to change the behavior, right?

**Comment by jacobmischka** [comment]
Sure, no reason not to

**Comment by felixsanz** [comment]
I also had problems with this until i came here. So bad UX

**Comment by limi58** [comment]
it is bad for chinese users, because pressing shift will switch english to chinese, it is so trouble.

**Comment by staticfox** [comment]
Also had to come to this issue to figure out how to scroll tabs :/

**Comment by Salamandar** [comment]
Same as @staticfox. It's just not natural to use shit+scroll :D
Besides, scrolling changes the tab (i.e the buffer) but when a non-visible tab is used, the tab strip does not scroll to make it visible : I am editing a file but the corresponding tab is invisible. Weird.

**Comment by madprops** [comment]
Maybe there shouldn't be scrolling at all. I like how Chrome and Sublime shrink the tabs, gives you an idea of where everything is even if it's kinda hard to read.

**Comment by Salamandar** [comment]
Sublime and Chrome allow scrolling. When you have 20+ tabs opened and you can't show the text for all of them (only like 15), you can scroll.

**Comment by qooob** [comment]
Hi, guys.

I must agree that using `shift+scroll` is actually very natural and intuitive for me (using Linux). However, it's surely not cool not to have tab automatically _PUT IN VIEW_ when it's focused. At the moment, even if you use search (Ctrl+T) and focus the tab that is not in view (many-many tabs and it's hidden - requires `shift+scroll`) - you don't get this focused (you don't see its name and it's not selected visually). Same if you `Ctrl+PageUp/PageDown`.

My point, please make optional to `bring tab to focus` upon making it active. And yes, Sublime does it this way.

Thank you Atom, switching from Sublime! ;) (after 5 years)

**Comment by Ben3eeE** [comment]
@qooob The behavior you describe is available in Atom 1.12. Not as an option but as the default behavior. When a new tab is activated it is automatically scrolled into view. You can look at the PR https://github.com/atom/tabs/pull/375 for more information.

**Comment by qooob** [comment]
Oh, great. Thanks. I'm on 1.11. Is the release expected soon? Is it safe to use betas?

**Comment by braydenhouston** [comment]
Just switched to Atom from Sublime a few weeks ago! Loving *almost* everything about the switch so far. This sadly has been the one imperfection. Couldn't find out how to scroll through tab until I found this thread. It would be nice to have the option to scroll through tabs without holding shift. Has there been any new development trying to make this an option?

**Comment by 50Wliu** [comment]
No, though we'd definitely appreciate and accept a well-crafted PR adding that functionality.

**Comment by pushamaku** [comment]
+1

**Comment by mcanepa** [comment]
ok, it's 2018...I hope this gets into release soon

**Comment by vertigo220** [comment]
I'll add my .02 by agreeing with the majority opinion here. In the past couple days, I've tried out several different editors in the search for a possible replacement to Notepad++. My initial impression of Atom was very good, but I didn't get very far at all into testing it before I abandoned it and moved on to the others due to this issue. I figured it was just poorly written and buggy, and after digging for a while looking to see if there was an option I was missing, I gave up. Only after trying and retrying all other editors, and finding them all to be unsatisfactory, did I return to Atom to try again, this time doing a search here and finding this issue. All the other editors (N++, Visual Studio Code, Sublime, Kate, etc) scroll the tabs when scrolling the (unmodified) mouse wheel. Every other program I've ever used that has tabs has worked this way as well. Some people may think the shift modifier to make it work is intuitive, but just because it is for _you_ doesn't mean it is for _most people_. All you have to do is look at the number of people that have posted here about it--and you can safely assume they represent a tiny fraction of the total number of people that have had issues due to it--to realize this is a common problem and most certainly _not_ intuitive to many people.

At the very least, this should be an option. Even better would be to allow the user to select not only whether or not shift is required, but whether it scrolls the tab bar or scrolls through the tabs, ideally with the ability to have it perform one of those actions with just scrolling and the other with a modifier. And finally, to avoid confusion (making shift+scroll the default even if there is an option to switch it might cause people to give up on it out of confusion and frustration), it wouldn't be a bad idea to pop up asking for the preferred behavior on first run. In any event, it seems that based on this and the possibility of data loss due to a longstanding and still unfixed bug, as nice as Atom appears to be it's not to the point I'd want to switch to it. It's disappointing that after 3.5 years such a simple issue hasn't been dealt with yet.

**Comment by Grant10k** [comment]
It took me for-ever to figure this out. I didn't even really know what to search for to find it. Finally "atom scroll tabs" and this result is like, the fifth item down as a bug report. It's obvious, but only in hindsight. 

I tried middle-click dragging the tabs (which lets you scroll around any direction on web pages and in another axis on some 3D programs), I tried right-click dragging, shift, ctrl, alt dragging. I tried like, looking for little tiny arrows on the edge. Finally I just had to just open the file again to reselect the tab I knew was open somewhere or made the atom window stretch across 2 monitors, select the tab I want, then shrink it back. 

Shift+Scroll did not even register as a possibility. 

For anyone else who looking to find their long lost tabs, mouses with left/right scroll wheel functionality also work if the mouse is over the tabs. I don't know if I tried that while I was trying everything, but at the time my mouse scroll left/right was bound to Home/End so it wouldn't worked even if I had thought to try it, but that's on me. That said, left/right scroll goes a little far in my opinion. I don't know if it's my specific mouse, but it seem to scroll to either end instantly, so it's impossible to get to middle tabs if you have an absurd amount open. 

In any case, beyond not requiring shift to scroll sideways, I'd like to see some way to drag the 'tabs' around instead of dragging a single tab around. Like you're grabbing a filmstrip instead of reordering a frame. Shift+Click or MiddleMouseClick could drag 'them'. Might help with discoverability. Also even if no one uses it, adding an option like "Don't require shift+scroll to scroll tabs" let's people know that it's even possible, and what the default method is, even if they don't change it.

**Comment by 50Wliu** [comment]
A PR was recently merged that fixed the tab scrolling option. Given that there is now an option to enable tab scrolling without the <kbd>shift</kbd> modifier on all platforms, is there anyone opposed to closing this issue?

**Comment by Grant10k** [comment]
Is this #553? It doesn't look directly related. I'm looking but I don't see any other PR that has anything to do with scrolling tabs since before 2018

**Comment by 50Wliu** [comment]
Ah, yep! I was mistaken @Grant10k, thanks for the correction.

**Comment by Grant10k** [comment]
I found out how to get the tab bar to scroll without hitting shift. 

The reason you need shift in the first place (and this will be 'no duh' to anyone who already knew this) is because Atom is basically a webpage (kudos to making it look so desktop-y btw). There is no logic that requires shift, it's just the tab bar is a zone without a scroll bar, and that's how chrome scrolls sideways. 

So what I did was create a custom package, loads jQuery (because I don't know how to hook into the mousewheel with regular JavaScript), hooks a mousewheel event to the second .tab-bar (the Project pane counts as the first one), and scrolls by the same amount. It's not as smooth as the shift+scroll, and if you try to make it smooth it jitters around when you scroll too fast so I opted to make the scroll not-smooth, but scroll half as far so it's easier to keep track of where you are. If one were to make this a 'real' package, there could probably be a fancy animation where scrolling moves the goalposts and the tab bar just animates smoothly until it matches up, but this current hack works well enough for me. 

```
    jQuery( document ).ready( function($){
      $(".tab-bar:eq(1)").on('mousewheel', function(event){
        if(!event.shiftKey){
          $(".tab-bar:eq(1)")[0].scrollLeft += -(event.originalEvent.wheelDelta / 2);
          //$(".tab-bar:eq(1)")[0].scrollBy({left:-event.originalEvent.wheelDelta, top:0, behavior: 'smooth'});
        }
      });
    });
```

**Comment by mcanepa** [comment]
> ok, it's 2018...I hope this gets into release soon

That was me on both windows 7 and 10.

But now I can confirm that this feature is working on linux (debian 9). If I scroll over the tabs they are scrolled. The only thing that is kind of annoying is that when the scroll reaches the end, then it stars from the other end and the scroll continues. I would expect to get to the end and stop.
